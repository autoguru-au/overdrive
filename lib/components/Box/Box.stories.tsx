import { Meta, StoryObj } from '@storybook/react-vite';
import React from 'react';

import { Box } from './Box';

const meta: Meta<typeof Box> = {
	title: 'Primatives/Box',
	tags: ['polymorphic'],
	component: Box,
	args: {
		as: 'div',
		children: 'Hello box!',
		colour: 'primary',
		backgroundColour: 'primary',
		borderRadius: 'none',
		borderColour: 'gray',
		borderWidth: '2',
		display: 'inline-flex',
		margin: undefined,
		padding: '6',
		textAlign: undefined,
	},
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Standard: Story = {};

/**
 * This story demonstrates the responsive props for `padding`, using an array of values.
 */
export const ResponsiveProps: Story = {
	args: {
		children: 'Resize the viewport',
		padding: ['3', '6', '8'],
	},
};

/**
 * Demonstrates the `odComponent` and `testId` props that map to data attributes built into Box and a custom attribute
 */
export const DataAttributes: Story = {
	render: () => (
		<Box
			id="so-basic"
			odComponent="box-basic"
			testId="basically-perfect"
			data-custom-attribute="somewhat less basic"
		>
			The most basic box (or is it?)
		</Box>
	),
};

/**
 * Passing in a React element to `as` props to merge style props
 */
// export const ComponentAsProp: Story = {
// 	render: () => (
// 		<Box
// 			as={<Box as="a" href="#hello" />}
// 			backgroundColor="accent"
// 			borderColor="info"
// 			borderWidth="1"
// 			className="keep-my-custom-class-name"
// 			p="4"
// 		>
// 			Styled props merged with custom component
// 		</Box>
// 	),
// };
