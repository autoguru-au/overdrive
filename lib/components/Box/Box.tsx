import cx from 'clsx';
import React, { FunctionComponent } from 'react';
import styles from './style.scss';

export enum EVariant {
	default = 'default',
	four = 'four',
}

export interface IProps {
	variant?: EVariant;
	borderColour?: string;
	strokeWidth?: 1 | 4;
	distance?: 0 | 1 | 2 | 3 | 4 | 5;
	className?: string;
}

export const Box: FunctionComponent<IProps> = ({
	variant = EVariant.default,
	borderColour = 'gray-300',
	strokeWidth = 1,
	distance = 0,
	children,
	className = '',
}) => (
	<div
		className={cx([
			styles.root,
			styles[`variant--${variant}`],
			styles[`borderColour--${borderColour}`],
			styles[`strokeWidth--${strokeWidth}`],
			styles[`distance--${distance}`],
			className,
		])}>
		{children}
	</div>
);
