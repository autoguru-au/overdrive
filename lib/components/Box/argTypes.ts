import { ArgTypes } from '@storybook/react';
import { ComponentProps } from 'react';

import { defaultGamut } from '../../themes/base/tokens';
import { Tokens } from '../../themes/tokens';

import { Box } from './Box';

export const scaleOptions: Array<keyof Tokens['space']> = [
	'none',
	'1',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
];
const scaledProps: Array<keyof ComponentProps<typeof Box>> = [
	'padding',
	'paddingX',
	'paddingY',
	'paddingTop',
	'paddingRight',
	'paddingLeft',
	'margin',
	'marginX',
	'marginY',
	'marginTop',
	'marginRight',
	'marginLeft',
];

const boxShadowOptions: Array<keyof Tokens['elevation']> = [
	'none',
	'1',
	'2',
	'3',
	'4',
	'5',
];
const borderRadiusOptions: Array<keyof Tokens['border']['radius']> = [
	'none',
	'1',
	'min',
	'full',
	'pill',
];
const widthOptions: Array<ComponentProps<typeof Box>['width']> = [
	'full',
	'none',
];
const oderOptions: Array<ComponentProps<typeof Box>['order']> = [0, 1, 2];
export const boxArgTypes: Partial<ArgTypes<ComponentProps<typeof Box>>> = {
	backgroundColour: {
		options: Object.keys(defaultGamut),
		control: {
			type: 'select',
		},
	},
	boxShadow: {
		options: boxShadowOptions,
		control: {
			type: 'select',
		},
	},
	borderRadius: {
		options: borderRadiusOptions,
		control: {
			type: 'select',
		},
	},
	alignItems: {
		options: ['flexStart', 'center', 'flexEnd', 'stretch'],
		control: {
			type: 'select',
		},
	},
	justifyContent: {
		options: ['flexStart', 'center', 'flexEnd', 'spaceBetween'],
		control: {
			type: 'select',
		},
	},
	flexDirection: {
		options: ['row', 'rowReverse', 'column', 'columnReverse'],
		control: {
			type: 'select',
		},
	},
	flexShrink: {
		options: {
			default: undefined,
			'1': '1',
		},
	},
	pointerEvents: {
		options: {
			default: undefined,
			none: 'none',
		},
	},
	width: {
		options: widthOptions,
		defaultValue: null,
		control: {
			type: 'select',
		},
	},
	order: {
		options: oderOptions,
		defaultValue: null,
		control: {
			type: 'select',
		},
	},
	...scaledProps.reduce((argTypes, prop) => {
		argTypes[prop] = {
			options: scaleOptions,
			control: {
				type: 'select',
			},
		};
		return argTypes;
	}, {}),
};
