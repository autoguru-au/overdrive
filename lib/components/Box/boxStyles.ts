import clsx from 'clsx';
import type { ElementType } from 'react';

import { element } from '../../reset/reset.css';
import {
	sprinkles,
	sprinklesLegacyColours,
	sprinklesResponsive,
} from '../../styles/sprinkles.css';
import { filterPropsWithStyles } from '../../utils/sprinkles';

import type { AsPolyProp, CustomProps, StyleProps } from './';

export type BoxStylesProps<E extends ElementType = 'div'> = AsPolyProp<E> &
	Pick<CustomProps, 'className'> &
	StyleProps;

const borderColorProps = [
	'borderColor',
	'borderBottomColor',
	'borderLeftColor',
	'borderRightColor',
	'borderTopColor',
];
const borderStyleProps = [
	'borderStyle',
	'borderBottomStyle',
	'borderLeftStyle',
	'borderRightStyle',
	'borderTopStyle',
];
const borderWidthProps = [
	'borderWidth',
	'borderBottomWidth',
	'borderLeftWidth',
	'borderRightWidth',
	'borderTopWidth',
];

/**
 * A function to be used for the `className` prop, returns combined base reset styles for the specified HTML element
 * and concatenates classes together with `className` can be an array, object, or other `clsx`-supported structure
 */
export const boxStyles = <E extends ElementType = 'div'>({
	as,
	className,
	...props
}: BoxStylesProps<E>) => {
	const tagName: string = as?.toString() ?? 'div';
	const resets = tagName in element ? element[tagName] : '';

	const hasBorderStyle = borderStyleProps.some((key) => !!props[key]);
	const hasBorderColorOrWidth = [
		...borderColorProps,
		...borderWidthProps,
	].some((key) => !!props[key]);

	const {
		sprinklesProps,
		sprinklesResponsiveProps,
		sprinklesLegacyColourProps,
	} = filterPropsWithStyles(props);

	// a little bit of logic specific border properties for backwards compatability
	if (hasBorderColorOrWidth && !hasBorderStyle) {
		sprinklesProps['borderStyle'] = 'solid';
	}

	return clsx(
		resets,
		sprinkles(sprinklesProps),
		sprinklesResponsive(sprinklesResponsiveProps),
		sprinklesLegacyColours(sprinklesLegacyColourProps),
		className,
	);
};
