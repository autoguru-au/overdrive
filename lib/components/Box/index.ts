export { Box } from './Box';
export { useBoxStyles } from './useBoxStyles';
