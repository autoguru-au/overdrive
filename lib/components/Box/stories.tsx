import { select, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { Box, EBoxVariant, IBoxProps } from '.';

const baseProps = (): IBoxProps => ({
	distance: select('Distance', [0, 1, 2, 3], 1),
	variant: select('Variant', EBoxVariant, EBoxVariant.default),
	borderColour: text('Border Colour', 'gray-300'), // TODO: This needs to be programmatic
	strokeWidth: select('Stroke Width', [1, 4], 1),
});

storiesOf('Components|Box', module).add('default', () => (
	<Box
		{...baseProps()}
		children={<div style={{ width: 150, height: 150 }} />}
	/>
));
