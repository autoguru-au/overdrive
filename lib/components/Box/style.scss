@import '../../theme/tokens/facade';

.root {
	display: inline-block;
	border: 1px solid transparent;
}

.variant {
	&--default {
		border-radius: 0;
	}

	&--four {
		border-radius: var(--global--space--1);
	}
}

.borderColour {
	&--gray-300 {
		border-color: getColour(gray, 300);
	}
}

.strokeWidth {
	&--1 {
		border-width: 1px;
	}

	&--4 {
		border-width: var(--global--space--1);
	}
}

.distance {
	&--0 {
		box-shadow: none;
	}

	&--1 {
		box-shadow: 0 0 2px 0 rgba(black, 0.04), 0 2px 16px 0 rgba(black, 0.08);
	}

	&--2 {
		box-shadow: 0 0 12px 0 rgba(black, 0.04),
			0 12px 48px 0 rgba(black, 0.08);
	}

	&--3 {
		box-shadow: 0 0 32px 0 rgba(black, 0.08),
			0 24px 96px 0 rgba(black, 0.12);
	}
}
