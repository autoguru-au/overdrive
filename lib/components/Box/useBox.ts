import type { ClassValue as ClassName } from 'clsx';
import type {
	ComponentPropsWithRef,
	ElementType,
	JSX,
	PropsWithChildren,
} from 'react';

import { useDeepCompareMemo } from '../../hooks';
import type {
	Sprinkles,
	SprinklesResponsive,
	SprinklesLegacyColours,
} from '../../styles/sprinkles.css';
import { filterNonSprinklesProps } from '../../utils/sprinkles';

import { boxStyles } from './boxStyles';

/**
 * Adds: `as` | `children` | `className`. Use BoxBasedProps to help consistently define the base props for a component
 * directly implementing useBox as the type of UseBoxProps is complex to Pick from.
 */
export interface BoxBasedProps<E extends ElementType>
	extends PropsWithChildren {
	/**
	 * The HTML element
	 * @default 'div'
	 */
	as?: E;
	/**
	 * Accepts `string` and complex array or objects via `clsx`
	 */
	className?: ClassName;
}

/** All vanilla-extract sprinkles props */
export type StyleProps = Sprinkles &
	SprinklesResponsive &
	SprinklesLegacyColours;

export type PolymorphicBoxProps<
	E extends ElementType,
	Props = object,
> = BoxBasedProps<E> &
	Omit<ComponentPropsWithRef<E>, keyof Props | 'as' | 'className'> &
	Props;

/** Polymorphic box props that merge sprinkles style props and the HTML element props */
export type UseBoxProps<E extends ElementType = 'div'> = PolymorphicBoxProps<
	E,
	StyleProps
>;

const DEFAULT_TAG = 'div' as keyof JSX.IntrinsicElements;
const LIST_ITEM_TAG = 'li' as keyof JSX.IntrinsicElements;
const LIST_TAGS = ['ul', 'ol'] as ReadonlyArray<keyof JSX.IntrinsicElements>;

/**
 * The Overdrive component primitive to expose a flexible HTML element as a fully typesafe React component
 * that provides intrinsic props as well as style props from vanilla-extract sprinkles
 *
 * @returns `{ Component, componentProps, SemanticChild }` where SemanticChild is optionally passed back based on the
 * `as` prop
 */
export const useBox = <E extends ElementType = 'div'>({
	as,
	className,
	...props
}: UseBoxProps<E>) => {
	const Component = as ?? DEFAULT_TAG;

	// logic to promote semantic HTML and ensure a child tag is correct for the `as` prop
	const isList = LIST_TAGS.includes(Component as keyof JSX.IntrinsicElements);
	const SemanticChild = isList ? LIST_ITEM_TAG : undefined;

	const style = useDeepCompareMemo(
		() =>
			boxStyles({
				as, // boxStyles uses the 'as' prop to determine css resets based on tag name
				className,
				...props,
			}),
		[className, props],
	);

	const remainingProps = useDeepCompareMemo(
		() => filterNonSprinklesProps(props),
		[props],
	);

	return {
		Component,
		componentProps: {
			className: style,
			...remainingProps,
		},
		SemanticChild,
	};
};
