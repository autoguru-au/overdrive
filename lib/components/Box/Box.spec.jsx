import React from 'react';
import { mount, shallow } from 'enzyme';
import { Box } from './Box';
import { EBoxVariant } from '.';

describe('<Box />', () => {
	it('should not throw', () => expect(() => shallow(<Box />)).not.toThrow());

	it('should pass on className to dom element', () => {
		const box = mount(<Box className="box-class" />);
		expect(box.hasClass('box-class')).toBeTruthy();
	});

	it('should match snapshot for default Box', () => {
		expect(shallow(<Box />)).toMatchSnapshot();
	});

	it('should match snapshot for distance 1 with rounder corners', () => {
		expect(shallow(<Box variant={EBoxVariant.four} />)).toMatchSnapshot();
	});

	it('should match snapshot for distance 2', () => {
		expect(shallow(<Box distance={2} />)).toMatchSnapshot();
	});

	it('should match snapshot for distance 3', () => {
		expect(shallow(<Box distance={3} />)).toMatchSnapshot();
	});

	it('should apply distance 0 styles', () => {
		const box = shallow(<Box distance={0} />);
		expect(box.hasClass('distance--0')).toBeTruthy();
	});

	it('should apply distance 1 styles', () => {
		const box = shallow(<Box distance={1} />);
		expect(box.hasClass('distance--1')).toBeTruthy();
	});

	it('should apply distance 2 styles', () => {
		const box = shallow(<Box distance={2} />);
		expect(box.hasClass('distance--2')).toBeTruthy();
	});

	it('should apply distance 3 styles', () => {
		const box = shallow(<Box distance={3} />);
		expect(box.hasClass('distance--3')).toBeTruthy();
	});

	it('should apply passed stroke width styles', () => {
		const box = shallow(<Box strokeWidth={4} />);
		expect(box.hasClass('strokeWidth--4')).toBeTruthy();
	});
});
