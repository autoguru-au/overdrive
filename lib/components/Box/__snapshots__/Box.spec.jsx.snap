// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> should match snapshot for default Box 1`] = `
<div
  class="reset_block__6swfds3"
/>
`;
